File: app/models/sale.py
# app/models/sale.py

from datetime import datetime, timezone

from sqlalchemy import Column, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from utils.db import Base


class Sale(Base):
    __tablename__ = "sales"

    id = Column(Integer, primary_key=True, index=True)

    customer_id = Column(
        Integer,
        ForeignKey("customers.id"),
        nullable=False,
        index=True,
    )

    machine_id = Column(
        Integer,
        ForeignKey("machines.id"),
        nullable=False,
        unique=True,
        index=True,
    )

    asking_price = Column(Float, nullable=True)
    sale_price = Column(Float, nullable=False)

    status = Column(String, nullable=False, default="sold")

    sold_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )

    delivered_at = Column(
        DateTime(timezone=True),
        nullable=True,
    )

    notes = Column(Text, nullable=True)

    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )

    updated_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    customer = relationship(
        "Customer",
        back_populates="sales",
    )

    machine = relationship(
        "Machine",
        back_populates="sale",
    )

File: front3/src/types/sale.ts
// front3/src/types/sale.ts

import type { Customer } from "./customer";
import type { Machine } from "./machine";

export interface Sale {
  id: number;
  customer_id: number;
  machine_id: number;

  asking_price: number | null;
  sale_price: number;

  status: "sold" | "delivered" | "cancelled";

  sold_at: string | null;
  delivered_at: string | null;

  billing_address_line_1: string | null;
  billing_address_line_2: string | null;
  billing_city: string | null;
  billing_state: string | null;
  billing_postal_code: string | null;

  delivery_same_as_billing: boolean;

  delivery_address_line_1: string | null;
  delivery_address_line_2: string | null;
  delivery_city: string | null;
  delivery_state: string | null;
  delivery_postal_code: string | null;

  notes: string | null;

  created_at: string;
  updated_at: string;
}

export interface SaleDetail extends Sale {
  customer: Customer;
  machine: Machine;
}